### ISSUE TITLE
longer alt text for hotspot image connector and possibly other image interactions

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#573]
I've been demonstrating using the hotspot image connector page for more visual scenario interactions and although it works great it's less accessible without a full size text box for the text that currently works as the alt tag for the hotspot. e.g. currently it's the label field that is used as the alt text for each hotspot but it's not as simple as adding the longer text to the label field because that causes a problem in the tree and is also **less easy to edit in the current single line text box**.
I'd like to add an optional multi-line plain text box for this which if not empty is used instead of the label field for the alt text.
Any comments/reasons not to do this?
I'm not actually sure if there's a technical limit to the length of alt text? I've read recommendations to keep short and succinct but I think that's for general use in webpages rather than the intended purpose here e.g. to make hotspot image connector interactions accessible for screenreader users.
There might other places/page types where it's worth doing this too?

**Comment by JohnSmith-LT** [comment]
JAWS screen reader apparently splits the alt text into blocks of 125 characters which is why numerous websites (https://stackoverflow.com/questions/27150277/what-is-the-limit-of-character-to-use-in-alt-text-of-img-tag-in-html) suggest less a max of 125 characters but the specification does not define a limit.

I will add this in and you can test.

**Comment by JohnSmith-LT** [comment]
That's it there @ronm123 

I've just called it Alt Text and use it if it has been added and is not empty otherwise I use the Tooltip as before. Let me know if you think the wording of any of these should be changed. Here's how it looks:

![image](https://user-images.githubusercontent.com/3907362/69013623-2a62ce80-097a-11ea-8b5f-d59c86c7d8fc.png)

**Comment by ronm123** [comment]
@JohnSmith-LT from your screenshot this looks good.  I'm probably not going to get a chance to test until Tuesday now.

**Comment by ronm123** [comment]
@JohnSmith-LT I've been testing this and can see that if the alt text optional property is added for a particular hotspot it uses that and if it not for another hotspot it just uses the tooltip as before. I also tested toggling the option to show tooltips on and off and can see that picks up and shows the alt text if added and just the tooltip field if not. I also tested tts with NVDA and that reads all of the alt text too (even when tooltips are set not to show on hover) so all great to that point. 

However what I did notice in testing this is that when tabbing with a screenreader turned on it's missing whatever is the first hotspot in the tree. So if I have four hotspots labelled hotspot 1 to hotspot 4 in that order and tab around the page until it reaches the hotspots it doesn't read hotspot 1 and starts reading at hotspot 2. Tab again and it reads hotspot 3, tab again and it reads hotspot 4, tab again and keep tabbing around the navigation elements until it reaches the hotspots again and it still misses the first hotspot. I'm guessing this is the hotspot array starting at zero or something like that?

**Comment by JohnSmith-LT** [comment]
Thanks @ronm123 good that the alt-text/tooltips are working as you'd expect.

The tabbing issue is probably the tabindex not being set, perhaps the first hotspot hasn't been getting one or it's getting the same one as the one before, like 1,2,2,3,4,5 etc and so it's skipping. Didn't think anything that I did would have affected that logic but will take a look later and see what I can do.

**Comment by ronm123** [comment]
@JohnSmith-LT I'm not saying it's anything that has changed as a result of your commit but just an issue I noticed while testing. Obviously it is an issue we need to fix though and I'll test the same thing with other hotspot pages later.

**Comment by JohnSmith-LT** [comment]
@ronm123 sorry I know that's not what you meant. Anyway, it was the tabIndex being set to 1 which was shared with the page text paragraph. I've fixed now and hotspot 1 gets tabindex 2...

**Comment by ronm123** [comment]
@JohnSmith-LT great works correctly now. I tested for the same tab order issue with annotated diagram, drag and drop labelling and hotspot image and they don't seem to have the same problem. I think we do ideally need the same or similar optional alt text option for those image interaction pages too. I think hotspot image would be the same as hotspot connector but annotated diagram and drag and drop labelling are a bit more complicated and probably worth listing as separate issues so that we can all work out and discuss how best to make those more accessible whilst maintaining the nature of the interactions. I'll do some further testing and then add annotated diagram and drag and drop labelling as separate issues.

Shall we leave this issue open until the alt text option is added to hotspot image?

**Comment by JohnSmith-LT** [comment]
That's this updated on the Hotspot Image:

![image](https://user-images.githubusercontent.com/3907362/69173922-ae42c500-0af8-11ea-902a-e5fd05a24a6c.png)

I'm curious though because in the xwd file there is mention of a Hotspot Question but you can't add it anywhere? Has it been deprecated or is this a bug? Not sure i've ever seen it in action:

![image](https://user-images.githubusercontent.com/3907362/69174043-f530ba80-0af8-11ea-8e24-fdb790ac8a42.png)

Any ideas @ronm123 or @FayCross 

Feel free to close this @ronm123 if you are happy with it now... and open another issue for the other things you mentioned.

**Comment by ronm123** [comment]
@JohnSmith-LT I think that’s probably something that @torinfo removed because it was added as a new feature but wasn’t optional and was therefore breaking existing hotspot pages/projects.

**Comment by torinfo** [comment]
@JohnSmith-LT @ronm123 Yes, it completely broke existing hotspot pages. So I am all for adding the possibility to add questions, but in a manner that is compatible with existing functionality.

**Comment by torinfo** [comment]
@ronm123 @JohnSmith-LT I would like to propose to add commit https://github.com/thexerteproject/xerteonlinetoolkits/commit/009d0568f9f6e3cc52f3ab73ac86ca9a0e1fb55c to v3.8 now.

**Comment by ronm123** [comment]
@torinfo I would propose adding all of the commits @JohnSmith-LT has made these last few days.

**Comment by JohnSmith-LT** [comment]
Yes fine with me @torinfo . Did you see that there are 3 commits in this issue? 1 and 3 do the alt text for Hotspot Image and Hotspot Connector pages and 2 fixes a tabbing issue in the connector that misses the first hotspot...

**Comment by JohnSmith-LT** [comment]
Maybe cherrypick the best of the other ones. Not the audio one yet I feel as we may have to write our own plugin to do what we need and incorporate recording also... but definitely the MathJax and dialog box ones as they are bugfixes and possibly the bullet page although it might be best to wait on that too to see if Fay can answer some of the last questions I have asked since she logged the issue...

**Comment by JohnSmith-LT** [comment]
> @JohnSmith-LT @ronm123 Yes, it completely broke existing hotspot pages. So I am all for adding the possibility to add questions, but in a manner that is compatible with existing functionality.

If you have time @torinfo can you open a new issue for this and maybe an outline of quick mockup of what you want it to do as i'm curious and not really seeing how it would work...

**Comment by ronm123** [comment]
@JohnSmith-LT just checking the ALT text addition to the hotspot image page and I've noticed a difference when it comes to existing projects. I'm not sure if this is the age of the projects I'm testing with or something else and I checked the wizards in my local copy of develop to see if I could spot a difference and couldn't. I did however notice that my \src\Nottingham\wizards\en-GB\ folder has connectorHotspotImage.xwd and connectorHotspotImage_XOT.xwd  is that correct?

After clearing cache etc when I open existing projects that have a hotspot image connector and select one of the existing hotspots I see the optional Alt text so that's great if I wanted I could add alt text to one or more of those existing hotspots.

However if I select one of the hotspots in an existing hotspot image page (not connector) I don't see the optional Alt text option. I only see that on any new hotspots I add to that same interaction or if I add a new hotspot image page and then new hotspots to that.

I can't spot why that would be the case but I know there's something similar with the new flexible hotspot functionality where you have to either add a new hotspot to an existing interaction or a totally new page to have the option of flexible hotspots but then that doesn't explian why I see the Alt text option on existing connector hotspots but not on the standard hotspot image hotspots.

It's not the end of the world to have to add new hotspots to existing interactions if alt text is required but seems strange that this isn't the case for the connector version.

@torinfo do you know why this is the case?

**Comment by JohnSmith-LT** [comment]
@ronm123 I did notice that there was a_XOT version of the connector and one without that extension and when I looked at the connector page in the editor I could quickly see that the optional attributes married up to the _XOT version. I do not know why there are two different ones though...

As for why the alt-text option is not being added to existing hotspots but is to new ones I do not know. I didn't text any old ones as I didn't have any and it is quite strange behaviour.

Maybe @torinfo has an idea why that might happen but if not then I will investigate further - are you able @ronm123 to email me a LO that exibits this behaviour or share a link to one and i'll take the xml file and put it into my setup to see if I can reproduce?

**Comment by torinfo** [comment]
@ronm123 @JohnSmith-LT 
 - The fact that it doesn't work on the connector, is because of the fact that you added the property to the flexhotspot, and the flexhotspot is only used when a new hotspot is added. Not when an old hotspot is edited. Bit of a long technical story, but the hotspot and the flexhotspot is shared between more than one page, and they ALL need te be identical. So, I need to look into that, because now they are not. I'll fix that. The real solution is to be able to upgrade/convert sub-items to other items in newer installations and then make sure that all sub items are unique between pages.
 - The _XOT version is an inheritance from the page types from the old standalone Xerte application that Jonathan introduced and maintained. Most of the xwd were identical whther they were used in the standalone Xerte application or in XOT. For the few exceptions, two versions were created where the _XOT version is used for XOT.

**Comment by torinfo** [comment]
@JohnSmith-LT Please let me know which commits you're comfortable with to cherry-pick to 3.8. And I rather be carefull, and first commit a few to 3.8.4, and then later commit a few other ones to 3.8.5

**Comment by ronm123** [comment]
@torinfo just a quick correction - according to my tests it (new alt text option) shows as available with existing hotspot image connector hotspots but not hotspot image hotspots.

**Comment by ronm123** [comment]
@torinfo @JohnSmith-LT I would suggest the following commits:
97f78ce7379c360a51e06b9669106653ae3ab3a8
78787e776e2e8dcfec83084c94436bff255f84d6
1aabebf7a95baf84f0c1eb2742b5b3714e46217b
96313c6c4394161112accf90f3655e153f5aed4c
067f63ce7fa015e227bf2760e210eea930a2eac9
8b0990e84657d08e7cce5a5d3053550b91f0283e
384a612bb778e0a0d8a7a998a128022459ca577c
25a45df10873a1d8aaf9c47e1b9f6cfde8bed5ed
640833ae3fe4bdaad6faad5eef2c18d5c688c18d
43b2064427b100b750ddcba1a05a9fd8b12798f8
f0b1ec9611f6fabf5d3b1bfd07f79ddca215e5f3
009d0568f9f6e3cc52f3ab73ac86ca9a0e1fb55c

In other words everything but the audio button as John mentioned but to be honest I think that's very useful as it is too.

**Comment by ronm123** [comment]
Also the tiny commit by Fay commenting out debugger 7f9d1683dc26897ac60d230d149c9af482637d6e

**Comment by JohnSmith-LT** [comment]
Yeah to be honest I think the audio button is useful as it is

**Comment by JohnSmith-LT** [comment]
> @torinfo just a quick correction - according to my tests it (new alt text option) shows as available with existing hotspot image connector hotspots but **not hotspot image hotspots**.

@ronm123 do you mean only EXISTING hotspot image hotspots and newly ceated ones work?

**Comment by ronm123** [comment]
@JohnSmith-LT @torinfo 

To be very clear:

Existing project with a hotspot image connector page and a normal hotspot page.
Select the existing hotspots on the connector page, or add new ones and the alt text option is available for both.
Select the existing hotspots on the normal hotspot page and the alt text isn't available but add a new hotspot to that interaction and the new hotspot does show the alt text option but the existing hotspots do not.

From what @torinfo has said this is because the new alt text option is added to the flexhotspot section so only appears when a new hotspot is added. I'm not sure how the logic works because flexhotspots are available in the connector page too  but still only have one entry in the wizard for alttext. Perhaps the alttext entry in the normal hotspot image wizard just needs to be moved to a different place in the wizard?

Attached is a quick test file.

[new_alt_text_test_deployment.zip](https://github.com/thexerteproject/xerteonlinetoolkits/files/3874496/new_alt_text_test_deployment.zip)

**Comment by JohnSmith-LT** [comment]
@ronm123 ah right, I get this now. Leave that with me and i'll see if I can fix that.

**Comment by torinfo** [comment]
@JohnSmith-LT As long as you make sure that in all the xwd's the hotspot and the flexhotspot need to be consistent.

**Comment by JohnSmith-LT** [comment]
So i've taken a look but it seems that in Hotspot Image there is no <hotspot> decalaration and it's now just the <flexhotspot> for new nodes but it has upgradefrom="hotspot"...

Without a declaration for <hotspot> in the xwd how do you add alttext for existing nodes?

![image](https://user-images.githubusercontent.com/3907362/69338270-b31f8a00-0c5a-11ea-9c2b-c876022dc3d2.png)

Should/can there be a <hotspot> declaration in there for existing ones OR do we need to upgrade the node when/if it is loaded and saved out?

**Comment by JohnSmith-LT** [comment]
So @torinfo adding back in the <hotspot> definition (well copying the <flexhotspot> definition and renaming to <hotspot>) puts the Alt Text option back onto existing LOs containing an older style hotspot but is this the correct way do you think or could/does that break something??

![image](https://user-images.githubusercontent.com/3907362/69339853-25459e00-0c5e-11ea-9dd9-24bd62e0cbc2.png)

I haven't committed this yet.

**Comment by torinfo** [comment]
@JohnSmith-LT Yes, that is how it works. Now, the hotspot in for example annotatedDaigram NEEDS to be the same. It also needs to include the alttext node.

As far as I can tell, these are the xwd's that need to be checked:
annotatedDiagram
dragDropLabel

**Comment by JohnSmith-LT** [comment]
Thanks @torinfo I have started working on this but now I have a few other questions and things that I wanted to note, even if just for the record...

You said that all the <hotspot> nodes need to be the same and I get that for consistencey however when I went to do the annotatedDiagram and dragDropLabel the Alt-Text option was already there. I'm assuming that this is because the <hotspot> node was present in the Hotspot Image node but then my question is, if all of these <hotspot> nodes interact like that then should they only be in one place?

Secondly, the user interaction for the annotatedDiagram and dragDrop is obviously different than for hotspot so it raises the question of where should the alt-text go?

- dragDrop - i have assumed that it goes on the drag label rather than on the actual hotspot whch would/could give the game away
- interactiveDiagram - does it go on the hotspot or on the actual link that activates the hotspot or both?

**Comment by JohnSmith-LT** [comment]
I have just realised that the hotspots being shared means that you can drag a hotspot from one type of page to a different page and they work... this is going to mean that we can't do that with newly created <flexhotspot> ones ?

**Comment by ronm123** [comment]
@JohnSmith-LT I would say the alt text goes on the hotspots for both drag and drop labelling and annotated diagram. 

If we take the bones in the foot example then for drag and drop labelling at the moment a screen reader just says drop zone for 1, drop zone for 2 etc. So what you might add to the alt text is something that describes what a sighted user sees within that particular hotspot. e.g. you might have: toe bones area, the five long bones, the three bones behind the five long bones but not mention Phalanx, Metatarsus or Cunieforms. That said this get ever more complicated! :-( (See additional issue listed below*)

If we take the annotated diagram example of bones in the foot then when a screen reader user tabs around each individual hotspot label is spoken (but not highlighted) and then each of the text links are spoken. So the alt text should definitely go on the hotspots especially as at the moment if you hit return or space on either the text links or the hotspots the screen reader doesn't then read the explanation text and I can't find a keyboard key or combination to focus on and read that.

*Additional issue:
In testing this with nvda I've noticed another fundamental issue that arguably makes adding the alt text on something like drag and drop labelling completely pointless! This may just be my lack of knowledge of settings in nvda but definitely something we need to investigate further. e.g. without nvda active I can tab to a label and hit space, tab to the relevant target and hit space and the label is automatically moved to the target. This works with lots of other interactions too so all good and something we have often shouted about e.g. drag and drop interactions can be operated by keyboard alone. However if I turn nvda on and repeat tabbing to a label I hear "draggable item, clickable metatarsus" but if I hit space and tab to one of the drop zone and hit tab again the label isn't moved and this is the same with other interactions too. I'm hoping it's just a setting in nvda that I/we need to know about otherwise all the interactions where something is meant to be dragged don't actually work for a screen reader user.

**Comment by JohnSmith-LT** [comment]
Ok @ronm123 I think that i've now done all 4 page models and applied your comments correctly, can you please check when you get a spare minute...

As for the nvda issue, i'll take a look at that too but maybe we should close this if you are happy the alt text stuff is all ok now and open a new issue because as you say that will most likely affect a lot more of the interactions...

**Comment by ronm123** [comment]
@JohnSmith-LT just tested and the alt text options appear for existing as well as new hotspots now on all four page types. However I'm not sure what your changes cover and what they don't yet cover. So I'll itemise each page type as follows:

1. Hotspot connector 
all seems to work fine including screen reader use picking up the alt text.

2. Annotated diagram
all seems to work ok including screen reader use picking up the alt text. 

3. Hotspot image page
works ok with show tooltip checked but with that unchecked when you tab to each of the hotspots a screen reader reads "play template ID equals...." I think with show tooltip unchecked the tooltips or alt text if added shouldn't show on hover (and they don't) but they should still be picked up by a screen reader and I think were before these latest changes.

4. Drag and drop labelling
The alt text options are available but tabbing to the targets still just reads "dropzone for 1", "dropzone for 2" etc. Obviously there hasn't been so far an option to show/hide tooltips on this page but probably should be and in that case should be the same as hotspot image e.g. don't show on hove but should still be read by a screen reader.

**Comment by JohnSmith-LT** [comment]
@ronm123 not sure what is going on with the Hotspot image in step 3. The only changes that have been made to that file are:

`$hotspot.attr("title", this.getAttribute("name"));`

to 

```
if (this.getAttribute("alttext") != undefined && this.getAttribute("alttext") != "")
        $hotspot.attr("title", this.getAttribute("alttext"));
else
        $hotspot.attr("title", this.getAttribute("name"));
```

which can't account for anything that you are seeing, especially the "play template ID equals...." Can you have a look in the underlying code for those hotspots (right click Inspect) to see if the xml file has somehow become corrupt and the title attribute is actually being set wrongly? I think they must have shown on hover before and the only way to make them not show on hover it to remove the text from the title tag. Possibly in a different tag the screen reader will still read them. Also, the 4 different page models use 3 different html tags to render the hotspot (which may be treated differently by the screen reader). Alluse the title attribute but it might be worth looking into what the screen readers read in each of these instances

- Hotspot Connector - uses <area>
- Hotspot Image - uses <area>
- Annotated Diagram - uses <a>
- Drag and Drop - uses <div>

As for Drag and Drop, the logic for that one is WAY more complex than the other 3 and the tooltip shown is updated in a dozen different places (for whatever reason) so it's going to take me a bit of time until I get to grips with that one.  Also, it has two different Interactivity modes that I hadn't accounted for...

**Comment by FayCross** [comment]
Only just getting round to looking at all of these issues you've commented on recently @JohnSmith-LT . Do you mind if I rename these new optional properties? Fields for alt text (e.g. for images) generally haven't been labelled as that as I'm not sure it would be understood but all users. We just use 'description' for image alt texts but maybe 'tooltip description' would be better for these?

Also, in Ron's initial request he asked for a larger text field to be used & these are still single line. I think they probably should be changed to the larger fields as otherwise it's not obvious what distinguishes them from the use of the default ones.

In drag & drop I'm not sure the two interactivity modes are needed to be honest - the only difference I can see between them is that one shows more text when the label is dropped. We could just detect whether there is any text in that field & not show if it's blank.

I think the drag & drop tooltips change quite a bit as years ago when this was first tested on screen readers I found that the testers wanted to be told when they could drop labels. So, for example a hotspot tooltip would just give a description when you are on it & no label has been picked up. But if you have a label that it's possible to drop then it will change to also include a brief instruction saying you can drop the label here. I think that's how it worked anyway - haven't looked.

**Comment by FayCross** [comment]
@torinfo if you are adding some commits to v3.8 could you please include this https://github.com/thexerteproject/xerteonlinetoolkits/commit/90031d7b79de9d3490e5228b7fc195b91fff7201 
I added some instructions for using the search functionality in Bootstrap but it wasn't quite in time for first 3.8 release
Thanks

**Comment by ronm123** [comment]
@JohnSmith-LT here's a demo of what I get with show tooltip unchecked with the normal hotspot image page. I tested with an existing hotspot image page, with a re-imported copy and with a newly created hotspot image page so I don't think it's due to any kind of corruption and as you'll see in the recording it's directly linked to whether the show tooltip is on or off. Obviously I'm testing with NVDA and don't know that well enough to be sure of settings etc but with the recording you'll see what I get.

https://drive.google.com/file/d/1Goi2BF8luo29N7WBwpA1ThQFe5KNquES/view?usp=sharing

**Comment by JohnSmith-LT** [comment]
Ah, so it is the screen reaer trying to be helpful - when there is no title attribute it looks for the next thing it can read and so it reads the href attribute which in this case is just # 

If you get a minute can you try adding in this highlighted line (sorry we have locked down computers and I have requested temporary admin rights to install NVDA for testing but IT being their usual it will take weeks!)

![image](https://user-images.githubusercontent.com/3907362/69429088-ae2a0b80-0d2a-11ea-8ce2-01fe7b846087.png)

It may be that in other pages where we optionally include the title tag we need to always add it and make it blank in some cases to stop this.

**Comment by ronm123** [comment]
@JohnSmith-LT I tried that but it wasn't making any difference. However I then checked the logic through those conditions and shouldn't line 244 be checking that the altext isn't undefine rather than is undefined? e.g. cuurent line 244 is:
`if (this.getAttribute("alttext") == undefined && this.getAttribute("alttext") != "")`
if I change that to:
`if (this.getAttribute("alttext") != undefined && this.getAttribute("alttext") != "")`
then it works as it should and the title or alt text is read out whether show tooltips is checked or not.

Let me know if that's correct and I can commit that change if you wish?

Also are you able to run portable applications? e.g. I run NVDA from a folder on my portable drive and just start it from a desktop shortcut. The thig I always forget if I haven't used it for a while is the keyboard shortcut for stopping it e.g. Caps Lock + Q.

**Comment by JohnSmith-LT** [comment]
I did notice that @ronm123 and thought it was to do with the default being on or something but then I added that line i'm sure so yes it is a typo. Good catch. If you are happy then yes feel free to submit it.

I will try the portable application option to see if it works but we've all had our computers locked down dy default because of an audit outcome so I'd be surprised if that option is still open...

**Comment by ronm123** [comment]
@JohnSmith-LT sorry even though I did correctly find the cause/solution I referred to the wrong line in my last post! :-(

It was line 243 on 244 that was incorrect and I have committed that change now.

This has become a long discussion so for clarity I think the following are things that we perhaps still need to agree/implement:

- [ ] Change of name of the alt text fields: @FayCross suggested these should be changed to  'tooltip description'. I guess to make this is easy the text that shows in the editor could change but leave the attribute name as alttext?

- [ ] Change the text fields to non wysyswg textareas to make it easier to enter longer descriptions.

- [ ] Improve the visibility of the on focus border when tabbing

- [ ] Not sure what @FayCross means about "In drag & drop I'm not sure the two interactivity modes are needed to be honest"? But also the second point about drag and drop target instructions?

- [ ] and a new question from me - with some, if not all, of these image interactions should/could we also have an optional property per hotspot or per interaction that they trigger on hover rather than on click?

**Comment by JohnSmith-LT** [comment]
> @JohnSmith-LT sorry even though I did correctly find the cause/solution I referred to the wrong line in my last post! :-(
> 
> It was line 243 on 244 that was incorrect and I have committed that change now.

Ah, yes that's the one I saw and that was there originally before I started work on this. I wondered if this was like that on purpose so that that section would be done if the tooltip flag isn't present - i.e in older LOs because otherwise older LOs may lose functionality... not sure though.

**Comment by JohnSmith-LT** [comment]
First attempt @ronm123 at patching Drag and Drop for the new alt text - turned out to be easier than i expected which reduces my expectations of success! If you have a fully formed drag drop exercise then can you test when you get a moment and see whether the screenreader likes it and if any bits turn out odd?

I will get to your checklist tomorrow.

**Comment by ronm123** [comment]
@JohnSmith-LT tabbing to the targets with NVDA running now reads the title or alt text if added of those targets. But I assume your fix isn't yet attempting to resolve any other issues with drag and drop labelling?

At the moment tab to label, space, tab to target, space, works without NVDA running but not with NVDA running. However this applies to other page types too like categories and gap fill drag and drop etc. However I've just tested with ChromeVox instead of NVDA and with that running the tab, space functionality works as it's meant to. I don't know whether that means there's just a setting in NVDA to change to get this working but I'll try to find out. Likewise I don't know how JAWS would work or whether there is still a free trial of Jaws that could be used for testing so I'll investigate that too.

**Comment by ronm123** [comment]
@JohnSmith-LT @FayCross it seems, as the saying goes, a little knowledge is a bad thing! ;-)

I skim read some of the basic NVDA tutorial at http://www.accessibilitycentral.net/nvda%20tutorials%20learning%20the%20basics.html and I wasn't aware there is a browse mode and a focus mode. It seems Ins + space toggles between the modes. From my testing it seems the tab, space, tab, space functionality (with Xerte at least) requires focus mode. With NVDA in focus mode this works and provides what sounds like useful instructions re press space to select, selected, press space to drop etc.

The one thing that doesn't yet happen is that when the item is dropped the description appears on the left but the screenreader (neither NVDA or ChromeVox) automatically reads that. I guess it isn't given focus?

**Comment by JohnSmith-LT** [comment]
@ronm123 that's good news that at least the toolltips seem to be consistent in your testing. I haven't got to the other things but I plan to attempt them either later today or tomorrow as it would be great to drap a line under this one...

Regarding not reading the description, there are possibly a coupleof ways we could handle that:

- could we fire off a timer which after a certain time (perhaps we could crudely calculate that based on the number of words in the tooltip - there will be no way to detect that a screen reader has finished reading the alt text)
- or perhaps we add a message to the end of the description saying to tab or press space or similar and we manipulate the tabindex order or just focus on the description to that it is read next?

Working all these things into the interactions for every model to make it useable for everyone makes me think that there should a button that toggles between different accessibility modes but i'm not sure whether that is best practice or te done thing... would make our life easier perhaps...

**Comment by JohnSmith-LT** [comment]
Have committed some changes to address the first 2 points...

**Comment by ronm123** [comment]
@JohnSmith-LT sorry didn't have a chance to test this until this morning...

The change to tooltip description seems fine as well as change to non-wysiwyg.

However with the hotspot image page, which has the option to control whether tooltips are shown or not, that didn't seem to work any longer e.g. tooltips were shown whether that option was checked or not. I looked at the if statement on line 243 again and could be wrong but concluded it still didn't make sense e.g. it was saying if hs_showTooltip doesn't equal undefined OR hs_showTooltip doesn't equal false. If I changed it back to does equal undefined or doesn't equal false then it worked again as far as controlling whether tooltips are shown or not but then with show tooltips unchecked a screen reader wouldn't hear the title or alt text when tabbing to the hotspots.

I changed OR to && on line 243 but then added an else and repeated the check for optional tooltip description so now it works like this:

If the show tooltip optional property isn't added or if it is checked then either the hotspot title or the tooltip description is added as a title attribute e.g. shows on hover and on focus.
If the show tooltip isn't undefined and is false then either the hotspot title or the tooltip description is added as alt attribute e.g. the tooltip doesn't show on hover but does get picked up by a screenreader when tabbing to the hotspots. I think this is how it should work bu ended up duplicating the check for tooltip description. It works according to my testing but if there's a better way or if you or @FayCross disagree that this is how it should work please change or comment.

**Comment by JohnSmith-LT** [comment]
I thought for a moment that you had the same code in the if and the else part but 1 is doing the title attribute and 1 the alt attribute so it looks absolutely fine to me if you are happy that the options are working as the should... it could be made smaller but it would lose readability.

I'm going to leave the 3 remaining points for now until @FayCross has a chance to comment on them and then we can work them out...
